### PR TITLE
Support for multiple OAUTH providers

### PR DESCRIPTION
**Pull request by longhotsummer**
This updates the OAUTH support to handle multiple providers. The login endpoint changes to `/api/1/sessions/login/<provider>` and the callback URL changes similarly. There is built-in support for Facebook and Google, and it's very easy to add support for other providers.

NB: the callback URL entered into the Google Developer Console must be updated.

When the user clicks the login dialog they get a modal popup giving them a choice of login providers.

This doesn't yet include full support for Twitter, because I don't yet have a functioning Twitter OAUTH setup.

sign-in

### CODE REVIEW

File: aleph/oauth.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        return
+    me = provider.get('userinfo')
+    user_id = 'google:%s' % me.data.get('id')
+    role = Role.load_or_create(user_id, Role.USER, me.data.get('name'),
+                               email=me.data.get('email'))
+    session['roles'].append(role.id)
+    session['user'] = role.id
+
+
+@signals.handle_oauth_session.connect
```
Comment (longhotsummer) [inline aleph/oauth.py]: We need the equivalent code for Twitter.

File: aleph/views/sessions_api.py [hunk 2]
```diff
@@ -60,13 +62,23 @@ def status():
...
             'write': authz.collections(authz.WRITE)
         },
-        'login': url_for('.login'),
-        'logout': url_for('.logout')
+        'logout': url_for('.logout'),
+        'providers': providers,
     })
 
 
 @blueprint.route('/api/1/sessions/login')
```
Comment (longhotsummer) [inline aleph/views/sessions_api.py]: This isn't technically necessary, we could remove it.

**Comment by pudo** [comment]
Looks very interesting, thanks much for the PR. Two thoughts:

a) Would it be possible to make it so the selector is skipped if there's only one provider configured (background: in our deployment here at OCCRP, the app runs against an internal OAuth provider and we can't allow people to pick), and

b) should the login perhaps be a proper route (e.g. ``/login``) so that we can put #52 there? On that note: @danohuiginn -- do you think it's worth back-porting the OpenOil code for password-based logins, too?

**Comment by longhotsummer** [comment]
I've adjusted it so that if there's only on oauth provider, it goes there directly.

I now also use triggerLogin, rather than a new global method.

Regarding `/login`, I don't mind either way. I'm not sufficiently familiar with Angular to know about popping up modals based on routes.

**Comment by longhotsummer** [comment]
The tests are failing because the app is being re-initialised for each test, but without the previous app being torn down correctly. Any thoughts?

**Comment by longhotsummer** [comment]
We have passing tests! This is ready to merge from my side.

**Comment by longhotsummer** [comment]
Can we merge?

**Comment by pudo** [comment]
Thanks very much for this pull, looks excellent.

**Pull request merged**